wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,221 @@
+    2. Open chrome via an intent to access the local copy.
+    3. Execute the benchmark - the copy has been modified to trigger the start of the benchmark.
 
-    known working chrome version 80.0.3987.149
-    '''
+    known working chrome version 83.0.4103.106
+    """
 
     parameters = [
-        Parameter('speedometer_version', allowed_values=versions, kind=str, default='2.0',
-                  description='''
-                  The speedometer version to be used.
-                  ''')
+        Parameter(
+            "chrome_package",
+            allowed_values=package_names,
+            kind=str,
+            default="com.android.chrome",
+            description="""
+                  The app package for the browser that will be launched.
+                  """,
+        ),
     ]
 
-    requires_network = True
-
     def __init__(self, target, **kwargs):
         super(Speedometer, self).__init__(target, **kwargs)
-        self.gui.timeout = 1500
-        self.gui.uiauto_params['version'] = self.speedometer_version
+        self.archive_asset = os.path.join(
+            os.path.dirname(__file__), "speedometer_archive.tgz"
+        )
+
+    def setup(self, context):
+        super(Speedometer, self).setup(context)
+
+        self.document_root = os.path.join(context.output_directory, "document_root")
+
+        self.webserver_port = get_available_webserver_port()
+
+        logging.info(
+            "Using port {} for temporary webserver".format(self.webserver_port)
+        )
+
+        self.speedometer_url = "http://localhost:{}/Speedometer2.0/index.html".format(self.webserver_port)
+
+        # Host a copy of Speedometer locally
+        subprocess.check_output(
+            "tar -xf {} -C {}".format(self.archive_asset, context.output_directory),
+            shell=True,
+        )
+
+        webserver_cmd = "cd {} && python3 -mhttp.server {}".format(
+            self.document_root, self.webserver_port
+        )
+
+        with open(os.devnull, "w") as dev_null:
+            self.webserver = subprocess.Popen(
+                webserver_cmd, shell=True, stdout=dev_null, stderr=dev_null
+            )
+
+        subprocess.check_output(
+            "adb reverse tcp:{0} tcp:{0}".format(self.webserver_port), shell=True
+        )
+
+        # Clear the cache...
+        self.target.execute("pm clear {}".format(self.chrome_package, as_root=True))
+
+        # Launch the browser for the first time...
+        browser_launch_cmd = "am start -a android.intent.action.VIEW -d {} {}".format(
+            self.speedometer_url, self.chrome_package
+        )
+
+        self.target.execute(browser_launch_cmd)
+        time.sleep(1)
+        self.target.execute("am force-stop {}".format(self.chrome_package))
+        time.sleep(1)
+
+        # Pull the preferences file from the device, modify it, and push it back.
+        # This is done to bypass the 'first launch' screen of the browser.
```

Comment:
Just a thought, would it have been more robust / remove the root requirement to still use uiautomator for performing this part the setup stage? If I remember correctly as long as it had exited before the run stage started it didn't impact the score?

Reply:
As I recall, the 'uiautomator dump' would only actually see the score when executed as root anyway, so I think we need to require root always.